### PR TITLE
New Powermax alarm binding

### PR DESCRIPTION
**Pull request by lolodomo**
Signed-off-by: lolodomo lg.hc@free.fr

### CODE REVIEW

**Comment by lolodomo** [comment]
@watou : no changes compared to my previous PR except that now there is no conflicts.
Previous PR was closed.
This PR is IMHO ready for a final merge.

**Comment by lolodomo** [comment]
Jenkins job failed due to a problem in enocean.test. But I changed nothing relative to enocean !

File: bundles/binding/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/PowerMaxBinding.java [hunk 1]
```diff
@@ -0,0 +1,875 @@
...
+        autoSyncTime = false;
+        pinCode = null;
+
+        PowerMaxReceiveType.POWERLINK.setHandlerClass(PowerMaxPowerlinkMessage.class);
+
+        if (config == null) {
+            logger.warn("PowerMax alarm binding: missing configuration");
+            this.setProperlyConfigured(false);
+            throw new ConfigurationException(null, "missing configuration");
+        }
```
Comment (watou) [inline bundles/binding/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/PowerMaxBinding.java]: I thought the proper response to receiving a `null` config in this method was to do nothing at all in this method?  Could you verify?

File: bundles/binding/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/PowerMaxGenericBindingProvider.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+     * {@inheritDoc}
+     */
+    @Override
+    public void validateItemType(Item item, String bindingConfig) throws BindingConfigParseException {
+
+        String selector[] = StringUtils.split(StringUtils.trim(bindingConfig), ':');
+        try {
+            PowerMaxSelectorType.fromString(selector[0]);
+        } catch (IllegalArgumentException e) {
+            throw new BindingConfigParseException("item '" + item.getName() + "': invalid selector " + bindingConfig);
```
Comment (watou) [inline bundles/binding/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/PowerMaxGenericBindingProvider.java]: If the only thing one should do in a `validateItemType` override is validate the item type, then this check should instead happen in `processBindingConfiguration`.

File: bundles/binding/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/message/PowerMaxSendType.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            PowerMaxReceiveType.POWERMASTER, "POWERMASTER_ZONE_STAT2"),
+    POWERMASTER_ZONE_NAME(new byte[] { (byte) 0xB0, 0x01, 0x21, 0x02, 0x05, 0x00, 0x43 }, null,
+            PowerMaxReceiveType.POWERMASTER, "POWERMASTER_ZONE_NAME"),
+    POWERMASTER_ZONE_TYPE1(new byte[] { (byte) 0xB0, 0x01, 0x2D, 0x02, 0x05, 0x00, 0x43 }, null,
+            PowerMaxReceiveType.POWERMASTER, "POWERMASTER_ZONE_TYPE1");
+
+    private byte[] message;
+    private Integer paramPosition;
+    private PowerMaxReceiveType expectedResponseType;
+    private String description;
```
Comment (watou) [inline bundles/binding/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/message/PowerMaxSendType.java]: Side fact: every `enum` has a `name()` method, which is the string version of the enum, so the `private String description` member seems to be redundant.

File: bundles/binding/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/state/PowerMaxPanelType.java [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+     */
+    public static PowerMaxPanelType fromLabel(String label) {
+        for (PowerMaxPanelType panelType : PowerMaxPanelType.values()) {
+            if (panelType.getLabel().equalsIgnoreCase(label)) {
+                return panelType;
+            }
+        }
+
+        throw new IllegalArgumentException("Invalid label: " + label);
+    }
```
Comment (watou) [inline bundles/binding/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/state/PowerMaxPanelType.java]: Side note: every `enum` has a `valueOf(String)` static method that would return the instance, so if you named your enum members using the labels, this method wouldn't be needed.

**Comment by watou** [comment]
Once again, impressive work @lolodomo!  I only added a couple of comments here.  Please address accordingly and I think this is ready for merge.

**Comment by lolodomo** [comment]
@watou: any idea why I got this strange jenkins job error with no link with my commit ?

I will apply the last changes, considering your comments.

**Comment by watou** [comment]
> any idea why I got this strange jenkins job error with no link with my commit ?

No, and job 359 has scrolled off.  As you said last night, it's probably unrelated to your PR.  If your local Maven build works, it's probably no issue.

File: bundles/binding/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/message/PowerMaxSendType.java (hunk 3 above)
Comment (lolodomo) [inline bundles/binding/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/message/PowerMaxSendType.java]: Fixed for PowerMaxSendType and PowerMaxReceiveType too.

File: bundles/binding/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/state/PowerMaxPanelType.java (hunk 4 above)
Comment (lolodomo) [inline bundles/binding/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/state/PowerMaxPanelType.java]: I want to keep a label adapted to user, for example PowerMax+, that is with lowercase and + in this case.

File: bundles/binding/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/PowerMaxGenericBindingProvider.java (hunk 2 above)
Comment (lolodomo) [inline bundles/binding/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/PowerMaxGenericBindingProvider.java]: I suppressed validateItemType.

Comment (lolodomo) [inline bundles/binding/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/PowerMaxGenericBindingProvider.java]: I suppressed all the code in validateItemType.

**Comment by lolodomo** [comment]
@watou: PR updated.
I updated my branch with the recent changes and then rebased to have my commit at last one:
```
git checkout master
git pull origin master
git checkout powermax
git merge master (no conflict detected)
git rebase -i HEAD~2 (just move my commit at end)
git checkout master
git push origin master
git checkout powermax
git push origin powermax -f
```
Jenkins jobs have passed and my local maven build is ok too.
I hope that is the right process and that I can now switch to another task...

**Comment by lolodomo** [comment]
In GitHub, I can see for my master branch: "This branch is even with openhab:master".
But for my powermax branch: "This branch is 24 commits ahead, 43 commits behind openhab:master.". I don't really understand why I am not only 1 commit behind openhab master.

**Comment by watou** [comment]
I think the issue is `git merge master` when you might have meant to `git rebase origin/master`.  The PR now has 24 commits in it.  There is probably some git-fu to make a new branch off current master, cherry-pick the commit you want in it, and submit that as a PR.  Sorry I'm not expert enough to give you worthwhile advice on it.  Almost there, though!

**Comment by lolodomo** [comment]
@watou: I finally got it ! But I am amazed how git is complicated to use !!! The amount of time I lost with git to provide my binding is absolutely terrible.

You were right, my error was certainly to use "git merge master", I think I should have used "git rebase master". In fact, I finally did it a little differently. First I rebased my powermax branch to delete the previous merge; then I switch to my master branch and run "git rebase powermax"; Finally I pushed my master branch in the remote powermax branch running "git push origin master:powermax -f".

**Comment by watou** [comment]
Great work, @lolodomo!

**Pull request merged**